File: development/NET/samples/server/Console/VB/MyDaAddressSpaceElement.vb
Imports System
Imports System.Collections
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace Console
	Public Class MyDaAddressSpaceElement
		Inherits DaAddressSpaceElement

#Region "Fields"
		Dim m_qualityGood As EnumQuality
#End Region

#Region "Constructors"
		'------------------------------

		Public Sub New(ByVal anItemID As String, ByVal aName As String, ByVal anUserData As UInt32, ByVal anObjectHandle As UInt32, ByVal aParentHandle As UInt32)
			MyBase.New(anItemID, aName, anUserData, anObjectHandle, aParentHandle)
			m_qualityGood = [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD)
		End Sub

		Public Sub New()
		End Sub

		'--
#End Region

#Region "Private Attributes"
		'-------------------------------

		Private m_properties As New Hashtable
		Private m_Type As Byte = TYPE_UNKNOWN

		'--
#End Region

#Region "Public Constants"
		'-------------------------------

		Public Shared TYPE_UNKNOWN As Byte = 0
		Public Shared TYPE_NODEMATH As Byte = 128
		Public Shared TYPE_NODERANDOM As Byte = 64
		Public Shared TYPE_RANDOM_100MS As Byte = 1
		Public Shared TYPE_RANDOM_1S As Byte = 2
		Public Shared TYPE_SINE As Byte = 4
		Public Shared TYPE_ACCEPT As Byte = 8

		'--
#End Region

#Region "Public Attributes"
		'-------------------------------

		Public Overridable Property Type() As Byte
			Get
				Return m_Type
			End Get
			Set(ByVal Value As Byte)
				m_Type = Value
			End Set
		End Property

		'--
#End Region

#Region "Public Methods"
		'---------------------

		''' <summary>
		''' Get elements property value data
		''' </summary>

		Public Sub GetPropertyValue(ByVal aRequest As DaRequest)
			If aRequest.PropertyId = 101 Then

				If Type = MyDaAddressSpaceElement.TYPE_RANDOM_100MS Then
					aRequest.Value = New ValueQT("Random value changing every 100 ms", m_qualityGood, DateTime.Now)
					aRequest.Result = EnumResultCode.S_OK
				ElseIf Type = MyDaAddressSpaceElement.TYPE_RANDOM_1S Then
					aRequest.Value = New ValueQT("Random value changing every second", m_qualityGood, DateTime.Now)
					aRequest.Result = EnumResultCode.S_OK
				ElseIf Type = MyDaAddressSpaceElement.TYPE_NODERANDOM Then
					aRequest.Value = New ValueQT("Random values", m_qualityGood, DateTime.Now)
					aRequest.Result = EnumResultCode.S_OK
				End If
			ElseIf aRequest.PropertyId = 6020 Then

				aRequest.Value = New ValueQT(OpcServer.g_angle, m_qualityGood, DateTime.Now)
				aRequest.Result = EnumResultCode.S_OK
			Else

				If m_properties.ContainsKey(aRequest.PropertyId) Then

					Dim [property] As DaConstantProperty = CType(m_properties(aRequest.PropertyId), DaConstantProperty)
					If Not [property] Is Nothing Then

						aRequest.Value = [property].Value
						aRequest.Result = EnumResultCode.S_OK
					End If
				End If

			End If

		End Sub

		Public Overloads Overrides Function QueryProperties(ByRef aPropertyList As ArrayList) As Integer
			If m_properties.Count > 0 Then
				aPropertyList = New ArrayList
				aPropertyList.AddRange(m_properties.Values)
			Else
				aPropertyList = Nothing
			End If
			Return EnumResultCode.S_OK
		End Function


		Public Function AddProperty(ByVal aProperty As DaProperty) As Integer
			If Not aProperty Is Nothing Then
				m_properties.Add(aProperty.Id, aProperty)
				Return EnumResultCode.S_OK
			Else
				Return EnumResultCode.S_FALSE
			End If
		End Function

		'--
#End Region

	End Class
End Namespace

File: development/NET/samples/server/Console/CS/MyDaAddressSpaceElement.cs
using System;
using System.Collections;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace Console
{
	public class MyDaAddressSpaceElement : DaAddressSpaceElement
	{

		#region Constructors
		//------------------------------

		public MyDaAddressSpaceElement(
			string anItemID,
			string aName,
			uint anUserData,
			uint anObjectHandle,
			uint aParentHandle)
			:
			base(
			anItemID,
			aName,
			anUserData,
			anObjectHandle,
			aParentHandle)
		{
		}   //	end constructor

		public MyDaAddressSpaceElement()
		{
		}   //	end constructor

		//--
		#endregion

		#region Private Attributes
		//-------------------------------

		private Hashtable m_properties = new Hashtable();
		private byte m_Type = TYPE_UNKNOWN;

		//--
		#endregion

		#region Public Constants
		//-------------------------------

		static public byte TYPE_UNKNOWN = 0x0;
		static public byte TYPE_NODEMATH = 0x80;
		static public byte TYPE_NODERANDOM = 0x40;
		static public byte TYPE_RANDOM_100MS = 0x01;
		static public byte TYPE_RANDOM_1S = 0x02;
		static public byte TYPE_SINE = 0x04;
		static public byte TYPE_ACCEPT = 0x08;

		//--
		#endregion

		#region Public Attributes
		//-------------------------------

		public virtual byte Type
		{
			get { return m_Type; }
			set { m_Type = value; }
		}   //	Attribute Type

		//--
		#endregion

		#region Public Methods
		//---------------------

		/// <summary>
		/// Get elements property value data
		/// </summary>

		public void GetPropertyValue(DaRequest aRequest)
		{
			if (aRequest.PropertyId == 101)
			{

				if (Type == MyDaAddressSpaceElement.TYPE_RANDOM_100MS)
				{
					aRequest.Value = new ValueQT("Random value changing every 100 ms", EnumQuality.GOOD, DateTime.Now);
					aRequest.Result = EnumResultCode.S_OK;
				}
				else if (Type == MyDaAddressSpaceElement.TYPE_RANDOM_1S)
				{
					aRequest.Value = new ValueQT("Random value changing every second", EnumQuality.GOOD, DateTime.Now);
					aRequest.Result = EnumResultCode.S_OK;
				}
				else if (Type == MyDaAddressSpaceElement.TYPE_NODERANDOM)
				{
					aRequest.Value = new ValueQT("Random values", EnumQuality.GOOD, DateTime.Now);
					aRequest.Result = EnumResultCode.S_OK;
				}   //	end if ... else
			}
			else if (aRequest.PropertyId == 6020)
			{

				aRequest.Value = new ValueQT(OpcServer.g_angle, EnumQuality.GOOD, DateTime.Now);
				aRequest.Result = EnumResultCode.S_OK;
			}
			else
			{

				if (m_properties.ContainsKey(aRequest.PropertyId))
				{

					DaConstantProperty property = m_properties[aRequest.PropertyId] as DaConstantProperty;
					if (property != null)
					{

						aRequest.Value = property.Value;
						aRequest.Result = EnumResultCode.S_OK;
					}   //	end if
				}   // end if

			}   //	end if ... else

		}   //	end GetPropertyValue

		public override int QueryProperties(out ArrayList aPropertyList)
		{
			if (m_properties.Count > 0)
			{
				aPropertyList = new ArrayList();
				aPropertyList.AddRange(m_properties.Values);
			}
			else
			{
				aPropertyList = null;
			}   //	end if ... else
			return (int)(EnumResultCode.S_OK);
		}   //	end QueryProperties


		public int AddProperty(DaProperty aProperty)
		{
			if (aProperty != null)
			{
				m_properties.Add(
					aProperty.Id,
					aProperty);
				return (int)EnumResultCode.S_OK;
			}   //	end if
			else
			{
				return (int)EnumResultCode.S_FALSE;
			}   //	end if...else
		}   //	end AddProperty

		//--
		#endregion

	}   //	end class MyAddressSpaceElement
}   //	end namespace
